hs_explore/models.py

```diff
@@ -0,0 +1,245 @@
+        self.save()
+
+    def unrelate(self, keyword):
+        """ unrelate a keyword to a recommended resource
+            :param keyword, the string keyword to be unrelated
+        """
+        updated_keyword_frequency = copy.deepcopy(self.keywords)
+        updated_keyword_frequency.pop(keyword, None)
+        self.keywords = updated_keyword_frequency
+        self.save()
+
+    def shown(self):
+        """ set the recommended resource status to shown """
+        self.state = Status.STATUS_SHOWN
+        self.save()
+
+    def explored(self):
+        """ set the recommended resource status to explored """
+        self.state = Status.STATUS_EXPLORED
+        self.save()
+        self.state = Status.STATUS_APPROVED
+        self.save()
+
+    def dismissed(self):
+        """ set the recommended resource status to dismissed """
+        self.state = Status.STATUS_DISMISSED
+        self.save()
+
+    @staticmethod
+    def delete(u, r):
+        """ delete a recommended resource record for a user
+            :param u, a User to delete a recommended resource
+            :param r, a Resource to be removed from the user's recommended resources
+        """
+        try:
+            record = RecommendedResource.objects.get(user=u, candidate_resource=r)
+            record.delete()
+        except Exception:
+            pass
+
+    @staticmethod
+    def clear():
+        """ clear all data records  """
+        RecommendedResource.objects.all().delete()
+
+    def to_json(self):
+        json_obj = {}
+        json_obj['user'] = self.user.username
+        json_obj['candidate_resource_id'] = self.candidate_resource.short_id
+        json_obj['candidate_resource_title'] = self.candidate_resource.title
+        json_obj['relevance'] = self.relevance
+        json_obj['rec_type'] = self.rec_type
+        json_obj['state'] = self.state
+        keywords = []
+        for keyword in list(self.keywords.items()):
+            keywords.append(keyword.to_json())
+        json_obj['keywords'] = keywords
+        return json_obj
+
+
+class UserPreferences(models.Model):
+    """ store user preferences to each keyword and resoruces that selected by the user """
+    user = models.OneToOneField(User, editable=False)
+    preferences = HStoreField(default={})
+    pref_for = models.CharField(max_length=8, choices=(('Resource', 'Resource'),
+                                ('User', 'User'), ('Group', 'Group')))
+
+    @staticmethod
+    def prefer(u, pf, preferences={}):
+        """ build up a user's preferences profile
+            :param u, a User to be set up the preferences
+            :param pf, choices from pref_for that describes the preferences for
+            :param preferences, a {keyword: frequency} dictionary represents the given
+            user's preferences for each keyword
+        """
+        defaults = {}
+        if len(preferences) > 0:
+            defaults['preferences'] = copy.deepcopy(preferences)
```

Comment:
is deepcopy actually needed?  preferences ends up in the db, which is effectually a copy

Reply:
Based upon my previous tests, it looks like we need to do a deepcopy.

Reply:
I can't think of a reason why deepcopy would be needed. @alvacouch - do you know why deepcopy would be necessary?